### PR TITLE
Few fixes for ART

### PR DESCRIPTION
**Pull request by muhzii**
- Fix hooking of Interpreter internals on Android 5.0 & 5.1.
- Improve the test suite to include running with optimizations enabled (in-lining optimizations disabled).
- Fail explicitly when hooking an ArtMethod if it's too short and the prologue can't be overwritten, instead of hanging.

### CODE REVIEW

File: lib/android.js [hunk 1]
```diff
@@ -2076,11 +2105,21 @@ class ArtQuickCodeInterceptor {
...
+
   activate (vm) {
     this._createTrampoline();
 
     const writeTrampoline = artQuickCodeReplacementTrampolineWriters[Process.arch];
     const prologueLength = writeTrampoline(this.trampoline, this.quickCode, vm);
+    if (prologueLength === 0) {
+      this._destroyTrampoline();
+
+      throw new Error('Unable to intercept quick code of ArtMethod.');
```
Comment (oleavr) [inline lib/android.js:2120]:
```suggestion
      throw new Error('Unable to intercept quick code of ArtMethod');
```

File: test/Makefile [hunk 2]
```diff
@@ -27,7 +27,11 @@ deploy: build/$(ANDROID_ABI)/runner build/tests.dex build/frida-java-bridge.js
 	adb push $^ build/$(ANDROID_ABI)/libartpalette.so $(deploy_data_dir)
 
 run:
-	adb shell "LD_LIBRARY_PATH='$(APEX_LIBDIRS):$(deploy_data_dir)' $(deploy_data_dir)/runner"
+	@echo
+	@echo "[*] Running test suite with optimizations enabled."
+	@adb shell "LD_LIBRARY_PATH='$(APEX_LIBDIRS):$(deploy_data_dir)' $(deploy_data_dir)/runner --enable-optimizations"
+	@echo "[*] Running test suite with optimizations disabled (interpreter mode)."
+	@adb shell "LD_LIBRARY_PATH='$(APEX_LIBDIRS):$(deploy_data_dir)' $(deploy_data_dir)/runner"
```
Comment (oleavr) [inline test/Makefile:34]: 🎉

File: test/runner.c [hunk 3]
```diff
@@ -93,7 +94,17 @@ main (int argc, char * argv[])
 
   js_backend = gum_script_backend_obtain_qjs ();
 
-  frida_java_init_vm (&vm, &env);
+  if (argc > 1 && strcmp(argv[1], "--enable-optimizations")) {
```
Comment (oleavr) [inline test/runner.c:97]:
```suggestion
  if (argc > 1 && strcmp (argv[1], "--enable-optimizations") == 0) {
```

File: test/runner.c [hunk 4]
```diff
@@ -83,6 +83,7 @@ main (int argc, char * argv[])
   jobjectArray argv_value;
   jstring data_dir_value, cache_dir_value;
   guint arg_index;
+  gboolean optimization_enabled = false;
```
Comment (oleavr) [inline test/runner.c:86]:
```suggestion
  gboolean optimization_enabled = FALSE;
```

(As it's a `gboolean` and not a `bool`.)

File: test/runner.c [hunk 5]
```diff
@@ -93,7 +94,17 @@ main (int argc, char * argv[])
 
   js_backend = gum_script_backend_obtain_qjs ();
 
-  frida_java_init_vm (&vm, &env);
+  if (argc > 1 && strcmp(argv[1], "--enable-optimizations")) {
+    optimization_enabled = true;
```
Comment (oleavr) [inline test/runner.c:98]:
```suggestion
    optimization_enabled = TRUE;
```

File: test/runner.c [hunk 6]
```diff
@@ -93,7 +94,17 @@ main (int argc, char * argv[])
...
+
+    // Pop argv[1].
```
Comment (oleavr) [inline test/runner.c:100]:
```suggestion
```

(It's clear from the code.)

File: test/runner.c [hunk 7]
```diff
@@ -93,7 +94,17 @@ main (int argc, char * argv[])
...
+    argv[1] = argv[0];
+
```
Comment (oleavr) [inline test/runner.c:102]:
```suggestion
```

File: test/runner.c [hunk 8]
```diff
@@ -142,11 +153,12 @@ main (int argc, char * argv[])
...
 
 static void
-frida_java_init_vm (JavaVM ** vm, JNIEnv ** env)
+frida_java_init_vm (JavaVM ** vm, JNIEnv ** env, gboolean optimization_enabled)
 {
   void * vm_module, * runtime_module;
   jint (* create_java_vm) (JavaVM ** vm, JNIEnv ** env, void * vm_args);
-  JavaVMOption options[6];
+  int n_options;
+  JavaVMOption *options;
```
Comment (oleavr) [inline test/runner.c:161]:
```suggestion
  JavaVMOption * options;
```

File: test/runner.c [hunk 9]
```diff
@@ -165,21 +177,40 @@ frida_java_init_vm (JavaVM ** vm, JNIEnv ** env)
   create_java_vm = dlsym (vm_module, "JNI_CreateJavaVM");
   g_assert (create_java_vm != NULL);
 
+  n_options = 5;
+  if (optimization_enabled) {
+    n_options += 4;
+  } else {
+    n_options += 1;
+  }
```
Comment (oleavr) [inline test/runner.c:185]:
```suggestion
  if (optimization_enabled)
    n_options += 4;
  else
    n_options += 1;
```

File: test/runner.c [hunk 10]
```diff
@@ -165,21 +177,40 @@ frida_java_init_vm (JavaVM ** vm, JNIEnv ** env)
...
+
+  options = malloc(n_options * sizeof(JavaVMOption));
```
Comment (oleavr) [inline test/runner.c:187]:
```suggestion
  options = g_new0 (JavaVMOption, n_options);
```

(Seems we were leaving `extraInfo` as random garbage, though I don't know if it's actually used.)

File: test/runner.c [hunk 11]
```diff
@@ -165,21 +177,40 @@ frida_java_init_vm (JavaVM ** vm, JNIEnv ** env)
...
+
   options[0].optionString = "-verbose:jni";
   options[1].optionString = "-verbose:gc";
   options[2].optionString = "-Xcheck:jni";
   options[3].optionString = "-Xdebug";
-  options[4].optionString = "-Xint";
-  options[5].optionString = "-Djava.class.path=" FRIDA_JAVA_TESTS_DATA_DIR "/tests.dex";
+  options[4].optionString = "-Djava.class.path=" FRIDA_JAVA_TESTS_DATA_DIR "/tests.dex";
+
+  if (optimization_enabled) {
```
Comment (oleavr) [inline test/runner.c:195]:
```suggestion
  if (optimization_enabled)
  {
```

File: test/runner.c [hunk 12]
```diff
@@ -165,21 +177,40 @@ frida_java_init_vm (JavaVM ** vm, JNIEnv ** env)
...
+    options[5].optionString = "-Xcompiler-option";
+    options[6].optionString = "--compiler-filter=speed";
+    options[7].optionString = "-Xcompiler-option";
+    options[8].optionString = "--inline-max-code-units=0";
+  } else {
```
Comment (oleavr) [inline test/runner.c:200]:
```suggestion
  }
  else
  {
```

File: test/runner.c [hunk 13]
```diff
@@ -165,21 +177,40 @@ frida_java_init_vm (JavaVM ** vm, JNIEnv ** env)
...
   args.version = JNI_VERSION_1_6;
-  args.nOptions = G_N_ELEMENTS (options);
+  args.nOptions = n_options;
   args.options = options;
   args.ignoreUnrecognized = JNI_TRUE;
 
   result = create_java_vm (vm, env, &args);
   g_assert_cmpint (result, ==, JNI_OK);
 
+  free(options);
```
Comment (oleavr) [inline test/runner.c:212]:
```suggestion
  g_free (options);
```

**Review by oleavr — CHANGES_REQUESTED** [review]
Thanks, this is awesome! 🎉

File: test/runner.c (hunk 3 above)
Comment (oleavr) [reply test/runner.c:97]: (Ah and the placement of braces needs to be consistent.)

File: test/runner.c (hunk 10 above)
Comment (muhzii) [reply test/runner.c]: The array is allocated differently depending on the mode of operation, though.

Comment (oleavr) [reply test/runner.c]: Yeah I meant that this was an existing bug, where we didn't initialize the `extraInfo` field of each option.

**Pull request merged**